python-lib/ontology_tagging/formatter_by_match.py

```diff
@@ -0,0 +1,165 @@
+        - self._duplicate_df is filled with the original rows of the Document Dataset, with copies
+        There are as many copies of a document as there are keywords in this document
+        Args:
+            row: pandas.Series from text_df
+            language_column: if not None, matcher will apply with the given language of the row
+        """
+        self.contains_match = False
+        language = self._get_document_language(row, language_column)
+        matches = []
+        document_to_match = self._get_document_to_match(row, language)
+        empty_row = {column: np.nan for column in self.tag_columns}
+        if not self.category_column:
+            matches = [
+                (
+                    self._matcher_dict[language](sentence, as_spans=True),
+                    row[self.text_column_tokenized][idx],
+                )
+                for idx, sentence in enumerate(document_to_match)
+            ]
+            self._get_tags_in_row(matches, row, language)
+        else:
+            self._get_tags_in_row_category(document_to_match, row, language)
+        if not self.contains_match:
+            self.output_df = self.output_df.append(empty_row, ignore_index=True)
+            self._duplicate_df = self._duplicate_df.append(
+                pd.DataFrame([row]), ignore_index=True
+            )
+
+    def _get_tags_in_row(self, matches: List, row: pd.Series, language: AnyStr) -> None:
+        """
+        Called by _write_row
+        Create the list of new rows with infos about the tags and gives it to _update_df
+        """
+        values = []
+        for match, sentence in matches:
+            values = [
+                self._list_to_dict(
+                    [
+                        self._keyword_to_tag[language][
+                            get_span_text(span=keyword, lemmatize=self.lemmatization)
+                        ],
+                        keyword.text,
+                        sentence,
+                    ]
+                )
+                for keyword in match
+            ]  # new rows to append to the output dataframe
+            self._update_df(match, values, row)
+
+    def _get_tags_in_row_category(
+        self, document_to_match: List, row: pd.Series, language: AnyStr
+    ) -> None:
+        """
+        Called by _write_row_category
+        Create the list of new rows with infos about the tags and gives it to _update_df function
+        """
+        tag_rows = []
+        original_document = list(row[self.text_column_tokenized])
+        for idx, sentence in enumerate(document_to_match):
+            tag_rows = [
+                self._list_to_dict(
+                    [
+                        keyword.label_,
+                        keyword.ent_id_,
+                        keyword.text,
+                        original_document[idx],
+                    ]
+                )
+                for keyword in sentence.ents
+            ]
```

Comment:
Same remark as above
```suggestion
        # For FormatterByMatch with categories, the self.tag_columns are ordered as: [category, tag, keyword, sentence]
        for idx, sentence in matches:
            new_rows = [
                {
                    self.tag_columns[0]: keyword.label_,
                    self.tag_columns[1]: keyword.ent_id_,
                    self.tag_columns[2]: keyword.text,
                    self.tag_columns[3]: original_document[idx],                    
                }
                for keyword in sentence.ents
            ]
```

Reply:
Good idea, I did it here ec3514d